File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/TrackerConfiguration.java
package com.snowplowanalytics.snowplow.configuration;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.snowplowanalytics.snowplow.internal.tracker.Logger;
import com.snowplowanalytics.snowplow.internal.tracker.TrackerConfigurationInterface;
import com.snowplowanalytics.snowplow.tracker.DevicePlatform;
import com.snowplowanalytics.snowplow.tracker.LoggerDelegate;
import com.snowplowanalytics.snowplow.tracker.LogLevel;

import org.json.JSONObject;

/**
 * This class represents the configuration of the tracker and the core tracker properties.
 * The TrackerConfiguration can be used to setup the tracker behaviour indicating what should be
 * tracked in term of automatic tracking and contexts/entities to track with the events.
 */
public class TrackerConfiguration implements TrackerConfigurationInterface, Configuration {
    public final static String TAG = TrackerConfiguration.class.getSimpleName();

    /**
     * @see #appId(String)
     */
    @NonNull
    public String appId;

    /**
     * @see #devicePlatform(DevicePlatform)
     */
    @NonNull
    public DevicePlatform devicePlatform;

    /**
     * @see #base64encoding(boolean)
     */
    public boolean base64encoding;

    /**
     * @see #logLevel(LogLevel)
     */
    @NonNull
    public LogLevel logLevel;

    /**
     * @see #loggerDelegate(LoggerDelegate)
     */
    @Nullable
    public LoggerDelegate loggerDelegate;

    /**
     * @see #applicationContext(boolean)
     */
    public boolean applicationContext;
    /**
     * @see #platformContext(boolean) 
     */
    public boolean platformContext;
    /**
     * @see #geoLocationContext(boolean) 
     */
    public boolean geoLocationContext;
    /**
     * @see #sessionContext(boolean) 
     */
    public boolean sessionContext;
    /**
     * @see #deepLinkContext(boolean)
     */
    public boolean deepLinkContext;
    /**
     * @see #screenContext(boolean) 
     */
    public boolean screenContext;
    /**
     * @see #screenViewAutotracking(boolean) 
     */
    public boolean screenViewAutotracking;
    /**
     * @see #lifecycleAutotracking(boolean) 
     */
    public boolean lifecycleAutotracking;
    /**
     * @see #installAutotracking(boolean) 
     */
    public boolean installAutotracking;
    /**
     * @see #exceptionAutotracking(boolean) 
     */
    public boolean exceptionAutotracking;
    /**
     * @see #diagnosticAutotracking(boolean) 
     */
    public boolean diagnosticAutotracking;
    /**
     * @see #userAnonymisation(boolean)
     */
    public boolean userAnonymisation;
    /**
     * @see #trackerVersionSuffix(String)
     */
    @Nullable
    public String trackerVersionSuffix;

    // Getters and Setters

    @Override
    @NonNull
    public String getAppId() {
        return appId;
    }

    @Override
    public void setAppId(@NonNull String appId) {
        this.appId = appId;
    }

    @Override
    @NonNull
    public DevicePlatform getDevicePlatform() {
        return devicePlatform;
    }

    @Override
    public void setDevicePlatform(@NonNull DevicePlatform devicePlatform) {
        this.devicePlatform = devicePlatform;
    }

    @Override
    public boolean isBase64encoding() {
        return base64encoding;
    }

    @Override
    public void setBase64encoding(boolean base64encoding) {
        this.base64encoding = base64encoding;
    }

    @Override
    @NonNull
    public LogLevel getLogLevel() {
        return logLevel;
    }

    @Override
    public void setLogLevel(@NonNull LogLevel logLevel) {
        this.logLevel = logLevel;
    }

    @Override
    @Nullable
    public LoggerDelegate getLoggerDelegate() {
        return loggerDelegate;
    }

    @Override
    public void setLoggerDelegate(@Nullable LoggerDelegate loggerDelegate) {
        this.loggerDelegate = loggerDelegate;
    }

    @Override
    public boolean isApplicationContext() {
        return applicationContext;
    }

    @Override
    public void setApplicationContext(boolean applicationContext) {
        this.applicationContext = applicationContext;
    }

    @Override
    public boolean isPlatformContext() {
        return platformContext;
    }

    @Override
    public void setPlatformContext(boolean platformContext) {
        this.platformContext = platformContext;
    }

    @Override
    public boolean isGeoLocationContext() {
        return geoLocationContext;
    }

    @Override
    public void setGeoLocationContext(boolean geoLocationContext) {
        this.geoLocationContext = geoLocationContext;
    }

    @Override
    public boolean isSessionContext() {
        return sessionContext;
    }

    @Override
    public void setSessionContext(boolean sessionContext) {
        this.sessionContext = sessionContext;
    }

    @Override
    public boolean isDeepLinkContext() {
        return deepLinkContext;
    }

    @Override
    public void setDeepLinkContext(boolean deepLinkContext) {
        this.deepLinkContext = deepLinkContext;
    }

    @Override
    public boolean isScreenContext() {
        return screenContext;
    }

    @Override
    public void setScreenContext(boolean screenContext) {
        this.screenContext = screenContext;
    }

    @Override
    public boolean isScreenViewAutotracking() {
        return screenViewAutotracking;
    }

    @Override
    public void setScreenViewAutotracking(boolean screenViewAutotracking) {
        this.screenViewAutotracking = screenViewAutotracking;
    }

    @Override
    public boolean isLifecycleAutotracking() {
        return lifecycleAutotracking;
    }

    @Override
    public void setLifecycleAutotracking(boolean lifecycleAutotracking) {
        this.lifecycleAutotracking = lifecycleAutotracking;
    }

    @Override
    public boolean isInstallAutotracking() {
        return installAutotracking;
    }

    @Override
    public void setInstallAutotracking(boolean installAutotracking) {
        this.installAutotracking = installAutotracking;
    }

    @Override
    public boolean isExceptionAutotracking() {
        return exceptionAutotracking;
    }

    @Override
    public void setExceptionAutotracking(boolean exceptionAutotracking) {
        this.exceptionAutotracking = exceptionAutotracking;
    }

    @Override
    public boolean isDiagnosticAutotracking() {
        return diagnosticAutotracking;
    }

    @Override
    public void setDiagnosticAutotracking(boolean diagnosticAutotracking) {
        this.diagnosticAutotracking = diagnosticAutotracking;
    }

    @Override
    public boolean isUserAnonymisation() {
        return userAnonymisation;
    }

    @Override
    public void setUserAnonymisation(boolean userAnonymisation) {
        this.userAnonymisation = userAnonymisation;
    }

    @Override
    @Nullable
    public String getTrackerVersionSuffix() {
        return trackerVersionSuffix;
    }

    @Override
    public void setTrackerVersionSuffix(@Nullable String trackerVersionSuffix) {
        this.trackerVersionSuffix = trackerVersionSuffix;
    }

    // Constructors

    /**
     * It sets a default TrackerConfiguration.
     * Default values:
     *         devicePlatform = DevicePlatform.Mobile;
     *         base64encoding = true;
     *         logLevel = LogLevel.OFF;
     *         loggerDelegate = null;
     *         sessionContext = true;
     *         applicationContext = true;
     *         platformContext = true;
     *         geoLocationContext = false;
     *         screenContext = true;
     *         deepLinkContext = true;
     *         screenViewAutotracking = true;
     *         lifecycleAutotracking = false;
     *         installAutotracking = true;
     *         exceptionAutotracking = true;
     *         diagnosticAutotracking = false;
     *         userAnonymisation = false;
     * @param appId Identifier of the app.
     */
    public TrackerConfiguration(@NonNull String appId) {
        this.appId = appId;

        devicePlatform = DevicePlatform.Mobile;
        base64encoding = true;

        logLevel = LogLevel.OFF;
        loggerDelegate = null;

        sessionContext = true;
        applicationContext = true;
        platformContext = true;
        geoLocationContext = false;
        deepLinkContext = true;
        screenContext = true;
        screenViewAutotracking = true;
        lifecycleAutotracking = false;
        installAutotracking = true;
        exceptionAutotracking = true;
        diagnosticAutotracking = false;
        userAnonymisation = false;
    }

    // Builder methods

    /**
     * Identifer of the app.
     */
    @NonNull
    public TrackerConfiguration appId(@NonNull String appId) {
        this.appId = appId;
        return this;
    }

    /**
     * It sets the device platform the tracker is running on.
     */
    @NonNull
    public TrackerConfiguration devicePlatform(@NonNull DevicePlatform devicePlatform) {
        this.devicePlatform = devicePlatform;
        return this;
    }

    /**
     * It indicates whether the JSON data in the payload should be base64 encoded.
     */
    @NonNull
    public TrackerConfiguration base64encoding(boolean base64encoding) {
        this.base64encoding = base64encoding;
        return this;
    }

    /**
     * It sets the log level of tracker logs.
     */
    @NonNull
    public TrackerConfiguration logLevel(@NonNull LogLevel logLevel) {
        this.logLevel = logLevel;
        return this;
    }

    /**
     * It sets the logger delegate that receive logs from the tracker.
     */
    @NonNull
    public TrackerConfiguration loggerDelegate(@Nullable LoggerDelegate loggerDelegate) {
        this.loggerDelegate = loggerDelegate;
        return this;
    }

    /**
     * Whether application context is sent with all the tracked events.
     */
    @NonNull
    public TrackerConfiguration applicationContext(boolean applicationContext) {
        this.applicationContext = applicationContext;
        return this;
    }

    /**
     * Whether mobile/platform context is sent with all the tracked events.
     */
    @NonNull
    public TrackerConfiguration platformContext(boolean platformContext) {
        this.platformContext = platformContext;
        return this;
    }

    /**
     * Whether geo-location context is sent with all the tracked events.
     *
     * @apiNote Requires Location permissions as per the requirements of the various
     * Android versions. Otherwise the whole context is skipped.
     */
    @NonNull
    public TrackerConfiguration geoLocationContext(boolean geoLocationContext) {
        this.geoLocationContext = geoLocationContext;
        return this;
    }

    /**
     * Whether session context is sent with all the tracked events.
     */
    @NonNull
    public TrackerConfiguration sessionContext(boolean sessionContext) {
        this.sessionContext = sessionContext;
        return this;
    }

    /**
     * Whether deepLink context is sent with all the ScreenView events.
     */
    @NonNull
    public TrackerConfiguration deepLinkContext(boolean deepLinkContext) {
        this.deepLinkContext = deepLinkContext;
        return this;
    }

    /**
     * Whether screen context is sent with all the tracked events.
     */
    @NonNull
    public TrackerConfiguration screenContext(boolean screenContext) {
        this.screenContext = screenContext;
        return this;
    }

    /**
     * Whether enable automatic tracking of ScreenView events.
     */
    @NonNull
    public TrackerConfiguration screenViewAutotracking(boolean screenViewAutotracking) {
        this.screenViewAutotracking = screenViewAutotracking;
        return this;
    }

    /**
     * Whether enable automatic tracking of background and foreground transitions.
     * @apiNote It needs the Foreground library installed.
     */
    @NonNull
    public TrackerConfiguration lifecycleAutotracking(boolean lifecycleAutotracking) {
        this.lifecycleAutotracking = lifecycleAutotracking;
        return this;
    }

    /**
     * Whether enable automatic tracking of install event.
     */
    @NonNull
    public TrackerConfiguration installAutotracking(boolean installAutotracking) {
        this.installAutotracking = installAutotracking;
        return this;
    }

    /**
     * Whether enable crash reporting.
     */
    @NonNull
    public TrackerConfiguration exceptionAutotracking(boolean exceptionAutotracking) {
        this.exceptionAutotracking = exceptionAutotracking;
        return this;
    }

    /**
     * Whether enable diagnostic reporting.
     */
    @NonNull
    public TrackerConfiguration diagnosticAutotracking(boolean diagnosticAutotracking) {
        this.diagnosticAutotracking = diagnosticAutotracking;
        return this;
    }

    /**
     * Whether to anonymise client-side user identifiers in session (userId, previousSessionId), subject (userId, networkUserId, domainUserId, ipAddress) and platform context entities (IDFA)
     */
    @NonNull
    public TrackerConfiguration userAnonymisation(boolean userAnonymisation) {
        this.userAnonymisation = userAnonymisation;
        return this;
    }

    /**
     * Decorate the v_tracker field in the tracker protocol.
     * @note Do not use. Internal use only.
     */
    @NonNull
    public TrackerConfiguration trackerVersionSuffix(@Nullable String trackerVersionSuffix) {
        this.trackerVersionSuffix = trackerVersionSuffix;
        return this;
    }

    // Copyable

    @NonNull
    @Override
    public Configuration copy() {
        TrackerConfiguration copy = new TrackerConfiguration(appId);
        copy.devicePlatform = devicePlatform;
        copy.base64encoding = base64encoding;

        copy.logLevel = logLevel;
        copy.loggerDelegate = loggerDelegate;

        copy.sessionContext = sessionContext;
        copy.applicationContext = applicationContext;
        copy.platformContext = platformContext;
        copy.geoLocationContext = geoLocationContext;
        copy.screenContext = screenContext;
        copy.deepLinkContext = deepLinkContext;
        copy.screenViewAutotracking = screenViewAutotracking;
        copy.lifecycleAutotracking = lifecycleAutotracking;
        copy.installAutotracking = installAutotracking;
        copy.exceptionAutotracking = exceptionAutotracking;
        copy.diagnosticAutotracking = diagnosticAutotracking;
        copy.userAnonymisation = userAnonymisation;
        copy.trackerVersionSuffix = trackerVersionSuffix;
        return copy;
    }

    // JSON Formatter

    public TrackerConfiguration(@NonNull String appId, @NonNull JSONObject jsonObject) {
        this(jsonObject.optString("appId", appId));
        String val = jsonObject.optString("devicePlatform", DevicePlatform.Mobile.getValue());
        devicePlatform = DevicePlatform.getByValue(val);
        base64encoding = jsonObject.optBoolean("base64encoding", base64encoding);
        String log = jsonObject.optString("logLevel", LogLevel.OFF.name());
        try {
            logLevel = LogLevel.valueOf(log.toUpperCase());
        } catch (Exception e) {
            Logger.e(TAG, "Unable to decode `logLevel from remote configuration.");
        }
        sessionContext = jsonObject.optBoolean("sessionContext", sessionContext);
        applicationContext = jsonObject.optBoolean("applicationContext", applicationContext);
        platformContext = jsonObject.optBoolean("platformContext", platformContext);
        geoLocationContext = jsonObject.optBoolean("geoLocationContext", geoLocationContext);
        screenContext = jsonObject.optBoolean("screenContext", screenContext);
        deepLinkContext = jsonObject.optBoolean("deepLinkContext", deepLinkContext);
        screenViewAutotracking = jsonObject.optBoolean("screenViewAutotracking", screenViewAutotracking);
        lifecycleAutotracking = jsonObject.optBoolean("lifecycleAutotracking", lifecycleAutotracking);
        installAutotracking = jsonObject.optBoolean("installAutotracking", installAutotracking);
        exceptionAutotracking = jsonObject.optBoolean("exceptionAutotracking", exceptionAutotracking);
        diagnosticAutotracking = jsonObject.optBoolean("diagnosticAutotracking", diagnosticAutotracking);
        userAnonymisation = jsonObject.optBoolean("userAnonymisation", userAnonymisation);
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/TrackerConfiguration.kt
/*
 * Copyright (c) 2015-2023 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.snowplow.configuration

import com.snowplowanalytics.core.tracker.Logger
import com.snowplowanalytics.core.tracker.PlatformContext
import com.snowplowanalytics.core.tracker.TrackerConfigurationInterface
import com.snowplowanalytics.core.tracker.TrackerDefaults
import com.snowplowanalytics.snowplow.tracker.DevicePlatform
import com.snowplowanalytics.snowplow.tracker.LogLevel
import com.snowplowanalytics.snowplow.tracker.LoggerDelegate
import org.json.JSONObject
import java.util.*

/**
 * The [TrackerConfiguration] can be used to set up the tracker behaviour, including what should be
 * tracked in term of automatic tracking, and entities to track with the events. 
 * Tracker logging can also be configured here.
 *
 * Default values:
 *  - devicePlatform: [DevicePlatform.Mobile]
 *  - base64encoding: true
 *  - logLevel: [LogLevel.OFF]
 *  - loggerDelegate: null
 *  - sessionContext: true
 *  - applicationContext: true
 *  - platformContext: true
 *  - geoLocationContext: false
 *  - screenContext: true
 *  - deepLinkContext: true
 *  - screenViewAutotracking: true
 *  - screenEngagementAutotracking: true
 *  - lifecycleAutotracking: false
 *  - installAutotracking: true
 *  - exceptionAutotracking: true
 *  - diagnosticAutotracking: false
 *  - userAnonymisation: false
*/
open class TrackerConfiguration : TrackerConfigurationInterface, Configuration {

    /**
     * Identifier of the app.
     */
    private var _appId: String? = null
    override var appId: String
        get() = _appId ?: sourceConfig?.appId ?: ""
        set(value) { if (value.isNotEmpty()) { _appId = value } }

    /**
     *  Fallback configuration to read from in case requested values are not present in this configuration.
     */
    var sourceConfig: TrackerConfiguration? = null

    private var _isPaused: Boolean? = null
    internal var isPaused: Boolean
        get() = _isPaused ?: sourceConfig?.isPaused ?: false
        set(value) { _isPaused = value }

    private var _devicePlatform: DevicePlatform? = null
    override var devicePlatform: DevicePlatform
        get() = _devicePlatform ?: sourceConfig?.devicePlatform ?: TrackerDefaults.devicePlatform
        set(value) { _devicePlatform = value }

    private var _base64encoding: Boolean? = null
    override var base64encoding: Boolean
        get() = _base64encoding ?: sourceConfig?.base64encoding ?: TrackerDefaults.base64Encoded
        set(value) { _base64encoding = value }

    private var _logLevel: LogLevel? = null
    override var logLevel: LogLevel
        get() = _logLevel ?: sourceConfig?.logLevel ?: TrackerDefaults.logLevel
        set(value) { _logLevel = value }

    private var _loggerDelegate: LoggerDelegate? = null
    override var loggerDelegate: LoggerDelegate?
        get() = _loggerDelegate ?: sourceConfig?.loggerDelegate
        set(value) { _loggerDelegate = value }

    private var _sessionContext: Boolean? = null
    override var sessionContext: Boolean
        get() = _sessionContext ?: sourceConfig?.sessionContext ?: TrackerDefaults.sessionContext
        set(value) { _sessionContext = value }

    private var _applicationContext: Boolean? = null
    override var applicationContext: Boolean
        get() = _applicationContext ?: sourceConfig?.applicationContext ?: TrackerDefaults.applicationContext
        set(value) { _applicationContext = value }

    private var _platformContext: Boolean? = null
    override var platformContext: Boolean
        get() = _platformContext ?: sourceConfig?.platformContext ?: TrackerDefaults.platformContext
        set(value) { _platformContext = value }

    private var _geoLocationContext: Boolean? = null
    override var geoLocationContext: Boolean
        get() = _geoLocationContext ?: sourceConfig?.geoLocationContext ?: TrackerDefaults.geoLocationContext
        set(value) { _geoLocationContext = value }

    private var _deepLinkContext: Boolean? = null
    override var deepLinkContext: Boolean
        get() = _deepLinkContext ?: sourceConfig?.deepLinkContext ?: TrackerDefaults.deepLinkContext
        set(value) { _deepLinkContext = value }

    private var _screenContext: Boolean? = null
    override var screenContext: Boolean
        get() = _screenContext ?: sourceConfig?.screenContext ?: TrackerDefaults.screenContext
        set(value) { _screenContext = value }

    private var _screenViewAutotracking: Boolean? = null
    override var screenViewAutotracking: Boolean
        get() = _screenViewAutotracking ?: sourceConfig?.screenViewAutotracking ?: TrackerDefaults.screenViewAutotracking
        set(value) { _screenViewAutotracking = value }

    private var _screenEngagementAutotracking: Boolean? = null
    override var screenEngagementAutotracking: Boolean
        get() = _screenEngagementAutotracking ?: sourceConfig?.screenEngagementAutotracking ?: TrackerDefaults.screenEngagementAutotracking
        set(value) { _screenEngagementAutotracking = value }

    private var _lifecycleAutotracking: Boolean? = null
    override var lifecycleAutotracking: Boolean
        get() = _lifecycleAutotracking ?: sourceConfig?.lifecycleAutotracking ?: TrackerDefaults.lifecycleAutotracking
        set(value) { _lifecycleAutotracking = value }

    private var _installAutotracking: Boolean? = null
    override var installAutotracking: Boolean
        get() = _installAutotracking ?: sourceConfig?.installAutotracking ?: TrackerDefaults.installAutotracking
        set(value) { _installAutotracking = value }

    private var _exceptionAutotracking: Boolean? = null
    override var exceptionAutotracking: Boolean
        get() = _exceptionAutotracking ?: sourceConfig?.exceptionAutotracking ?: TrackerDefaults.exceptionAutotracking
        set(value) { _exceptionAutotracking = value }

    private var _diagnosticAutotracking: Boolean? = null
    override var diagnosticAutotracking: Boolean
        get() = _diagnosticAutotracking ?: sourceConfig?.diagnosticAutotracking ?: TrackerDefaults.diagnosticAutotracking
        set(value) { _diagnosticAutotracking = value }

    private var _userAnonymisation: Boolean? = null
    override var userAnonymisation: Boolean
        get() = _userAnonymisation ?: sourceConfig?.userAnonymisation ?: TrackerDefaults.userAnonymisation
        set(value) { _userAnonymisation = value }

    private var _trackerVersionSuffix: String? = null
    override var trackerVersionSuffix: String?
        get() = _trackerVersionSuffix ?: sourceConfig?.trackerVersionSuffix
        set(value) { _trackerVersionSuffix = value }

    private var _platformContextProperties: List<PlatformContextProperty>? = null
    /**
     * List of properties of the platform context to track.
     * If not passed and `platformContext` is enabled, all available properties will be tracked.
     * The required `osType`, `osVersion`, `deviceManufacturer`, and `deviceModel` properties will be tracked in the entity regardless of this setting.
     */
    open var platformContextProperties: List<PlatformContextProperty>?
        get() = _platformContextProperties ?: sourceConfig?.platformContextProperties
        set(value) { _platformContextProperties = value }

    // Builder methods
    
    /**
     * Identifier of the app. This will be included in every event.
     */
    fun appId(appId: String): TrackerConfiguration {
        this.appId = appId
        return this
    }

    /**
     * Specify the device platform the tracker is running on.
     */
    fun devicePlatform(devicePlatform: DevicePlatform): TrackerConfiguration {
        this.devicePlatform = devicePlatform
        return this
    }

    /**
     * It indicates whether the JSON data in the payload should be base64 encoded.
     */
    fun base64encoding(base64encoding: Boolean): TrackerConfiguration {
        this.base64encoding = base64encoding
        return this
    }

    /**
     * It sets the log level of tracker logs.
     */
    fun logLevel(logLevel: LogLevel): TrackerConfiguration {
        this.logLevel = logLevel
        return this
    }

    /**
     * It sets the logger delegate that receive logs from the tracker. Default is STDOUT.
     */
    fun loggerDelegate(loggerDelegate: LoggerDelegate?): TrackerConfiguration {
        this.loggerDelegate = loggerDelegate
        return this
    }

    /**
     * Whether the application context entity should be sent with all the tracked events.
     */
    fun applicationContext(applicationContext: Boolean): TrackerConfiguration {
        this.applicationContext = applicationContext
        return this
    }

    /**
     * Whether the mobile/platform context entity should be sent with all the tracked events.
     */
    fun platformContext(platformContext: Boolean): TrackerConfiguration {
        this.platformContext = platformContext
        return this
    }

    /**
     * Whether the geo-location context entity should be sent with all the tracked events. 
     * The location is based off the last cached location in the device.
     * Requires Location permissions as per the requirements of the various
     * Android versions. Otherwise no entity is added at all.
     */
    fun geoLocationContext(geoLocationContext: Boolean): TrackerConfiguration {
        this.geoLocationContext = geoLocationContext
        return this
    }

    /**
     * Whether the session context entity should be sent with all the tracked events.
     */
    fun sessionContext(sessionContext: Boolean): TrackerConfiguration {
        this.sessionContext = sessionContext
        return this
    }

    /**
     * Whether the [DeepLink](com.snowplowanalytics.snowplow.entity.DeepLink) context entity 
     * should be sent with the first ScreenView event following a deep link.
     */
    fun deepLinkContext(deepLinkContext: Boolean): TrackerConfiguration {
        this.deepLinkContext = deepLinkContext
        return this
    }

    /**
     * Whether the screen context entity should be sent with all the tracked events.
     */
    fun screenContext(screenContext: Boolean): TrackerConfiguration {
        this.screenContext = screenContext
        return this
    }

    /**
     * Whether to enable automatic tracking of ScreenView events. 
     * Note that this automatic tracking relies on the Android 
     * `Application.ActivityLifecycleCallbacks` interface, and therefore only works with Activity-based
     * apps. Composable screens or views in Jetpack Compose apps are not autotracked.
     */
    fun screenViewAutotracking(screenViewAutotracking: Boolean): TrackerConfiguration {
        this.screenViewAutotracking = screenViewAutotracking
        return this
    }

    /**
     * Whether enable tracking the screen end event and the screen summary context entity.
     * Make sure that you have lifecycle autotracking enabled for screen summary to have complete information.
     */
    fun screenEngagementAutotracking(screenEngagementAutotracking: Boolean): TrackerConfiguration {
        this.screenEngagementAutotracking = screenEngagementAutotracking
        return this
    }

    /**
     * Whether to enable automatic tracking of background and foreground transitions. 
     * The Foreground library must be installed.
     */
    fun lifecycleAutotracking(lifecycleAutotracking: Boolean): TrackerConfiguration {
        this.lifecycleAutotracking = lifecycleAutotracking
        return this
    }

    /**
     * Whether to enable automatic tracking of install event.
     * In case com.android.installreferrer:installreferrer library is present,
     * an entity with the referrer details will be attached to the install event.
     */
    fun installAutotracking(installAutotracking: Boolean): TrackerConfiguration {
        this.installAutotracking = installAutotracking
        return this
    }

    /**
     * Whether to enable crash reporting.
     */
    fun exceptionAutotracking(exceptionAutotracking: Boolean): TrackerConfiguration {
        this.exceptionAutotracking = exceptionAutotracking
        return this
    }

    /**
     * Whether to enable diagnostic reporting.
     */
    fun diagnosticAutotracking(diagnosticAutotracking: Boolean): TrackerConfiguration {
        this.diagnosticAutotracking = diagnosticAutotracking
        return this
    }

    /**
     * Whether to anonymise client-side user identifiers in session (userId, previousSessionId), 
     * subject (userId, networkUserId, domainUserId, ipAddress) and platform context entities (IDFA)
     */
    fun userAnonymisation(userAnonymisation: Boolean): TrackerConfiguration {
        this.userAnonymisation = userAnonymisation
        return this
    }

    /**
     * Do not use. Internal use only. 
     * 
     * Decorate the v_tracker field in the tracker protocol.
     * @suppress
     */
    fun trackerVersionSuffix(trackerVersionSuffix: String?): TrackerConfiguration {
        this.trackerVersionSuffix = trackerVersionSuffix
        return this
    }

    /**
     * List of properties of the platform context to track.
     * If not passed and `platformContext` is enabled, all available properties will be tracked.
     * The required `osType`, `osVersion`, `deviceManufacturer`, and `deviceModel` properties will be tracked in the entity regardless of this setting.
     */
    fun platformContextProperties(platformContextProperties: List<PlatformContextProperty>?): TrackerConfiguration {
        this.platformContextProperties = platformContextProperties
        return this
    }

    // Copyable
    override fun copy(): Configuration {
        return TrackerConfiguration(appId)
            .devicePlatform(devicePlatform)
            .base64encoding(base64encoding)
            .logLevel(logLevel)
            .loggerDelegate(loggerDelegate)
            .sessionContext(sessionContext)
            .applicationContext(applicationContext)
            .platformContext(platformContext)
            .geoLocationContext(geoLocationContext)
            .screenContext(screenContext)
            .deepLinkContext(deepLinkContext)
            .screenViewAutotracking(screenViewAutotracking)
            .screenEngagementAutotracking(screenEngagementAutotracking)
            .lifecycleAutotracking(lifecycleAutotracking)
            .installAutotracking(installAutotracking)
            .exceptionAutotracking(exceptionAutotracking)
            .diagnosticAutotracking(diagnosticAutotracking)
            .userAnonymisation(userAnonymisation)
            .trackerVersionSuffix(trackerVersionSuffix)
            .platformContextProperties(platformContextProperties)
    }

    /**
     * @param appId Identifier of the app.
     */
    constructor(appId: String) {
        this._appId = appId
    }

    /**
     * This constructor is only used internally in the service provider
     */
    internal constructor()

    // JSON Formatter
    /**
     * This constructor is used in remote configuration.
     */
    constructor(appId: String, jsonObject: JSONObject) : this(
        jsonObject.optString(
            "appId",
            appId
        )
    ) {
        val value = jsonObject.optString("devicePlatform", DevicePlatform.Mobile.value)
        _devicePlatform = DevicePlatform.getByValue(value)
        if (jsonObject.has("base64encoding")) { _base64encoding = jsonObject.getBoolean("base64encoding") }

        if (jsonObject.has("logLevel")) {
            val log = jsonObject.optString("logLevel", LogLevel.OFF.name)
            try {
                _logLevel = LogLevel.valueOf(log.uppercase(Locale.getDefault()))
            } catch (e: Exception) {
                Logger.e(TAG, "Unable to decode logLevel from remote configuration.")
            }
        }

        if (jsonObject.has("sessionContext")) { _sessionContext = jsonObject.getBoolean("sessionContext") }
        if (jsonObject.has("applicationContext")) { _applicationContext = jsonObject.getBoolean("applicationContext") }
        if (jsonObject.has("platformContext")) { _platformContext = jsonObject.getBoolean("platformContext") }
        if (jsonObject.has("geoLocationContext")) { _geoLocationContext = jsonObject.getBoolean("geoLocationContext") }
        if (jsonObject.has("screenContext")) { _screenContext = jsonObject.getBoolean("screenContext") }
        if (jsonObject.has("deepLinkContext")) { _deepLinkContext = jsonObject.getBoolean("deepLinkContext") }
        if (jsonObject.has("screenViewAutotracking")) { _screenViewAutotracking = jsonObject.getBoolean("screenViewAutotracking") }
        if (jsonObject.has("screenEngagementAutotracking")) { _screenEngagementAutotracking = jsonObject.getBoolean("screenEngagementAutotracking") }
        if (jsonObject.has("lifecycleAutotracking")) { _lifecycleAutotracking = jsonObject.getBoolean("lifecycleAutotracking") }
        if (jsonObject.has("installAutotracking")) { _installAutotracking = jsonObject.getBoolean("installAutotracking") }
        if (jsonObject.has("exceptionAutotracking")) { _exceptionAutotracking = jsonObject.getBoolean("exceptionAutotracking") }
        if (jsonObject.has("diagnosticAutotracking")) { _diagnosticAutotracking = jsonObject.getBoolean("diagnosticAutotracking") }
        if (jsonObject.has("userAnonymisation")) { _userAnonymisation = jsonObject.getBoolean("userAnonymisation") }
    }

    companion object {
        val TAG = TrackerConfiguration::class.java.simpleName
    }
}
